plugins/modules/proxmox.py

```diff
@@ -518,22 +756,35 @@ def update_config(self, vmid, node, disk, cpus, memory, swap, **kwargs):
                 )
 
         # Remove all empty kwarg entries
-        kwargs = dict((k, v) for k, v in kwargs.items() if v is not None)
+        kwargs = dict((key, val) for key, val in kwargs.items() if val is not None)
 
         if cpus is not None:
             kwargs["cpulimit"] = cpus
         if disk is not None:
-            kwargs["rootfs"] = disk
+            kwargs["disk_volume"] = parse_disk_string(disk)
+        if "disk_volume" in kwargs:
+            if not all(isinstance(val, str) for val in kwargs["disk_volume"].values()):
+                self.module.warn("All disk_volume values must be strings. Converting non-string values to strings.")
+                kwargs["disk_volume"] = {key: str(val) for key, val in kwargs["disk_volume"].items()}
+            disk_dict = build_volume(key="rootfs", **kwargs.pop("disk_volume"))
+            kwargs.update(disk_dict)
         if memory is not None:
             kwargs["memory"] = memory
         if swap is not None:
             kwargs["swap"] = swap
         if "netif" in kwargs:
-            kwargs.update(kwargs["netif"])
-            del kwargs["netif"]
+            kwargs.update(kwargs.pop("netif"))
         if "mounts" in kwargs:
-            kwargs.update(kwargs["mounts"])
-            del kwargs["mounts"]
+            kwargs["mount_volumes"] = convert_mounts(kwargs.pop("mounts"))
+        if "mount_volumes" in kwargs:
+            mounts_list = kwargs.pop("mount_volumes")
+            for mount_config in mounts_list:
+                if not all(isinstance(val, str) for val in mount_config.values()):
+                    self.module.warn("All mount_volumes values must be strings. Converting non-string values to strings.")
+                    mount_config = {key: str(val) for key, val in mount_config.items()}
```

Comment:
Better use `string_types` instead of `str` (`from ansible.module_utils.six import string_types`), and use `to_text()` instead of `str()` (`from ansible.module_utils.common.text.converters import to_text`).
```suggestion
                if not all(isinstance(val, string_types) for val in mount_config.values()):
                    self.module.warn("All mount_volumes values must be strings. Converting non-string values to strings.")
                    mount_config = {key: to_text(val) for key, val in mount_config.items()}
```
Same further below.